Refacto provider and add Sources

gchoqueux opened the pull request:
## Description
add `Source` in layer : all the parameters to get the data from a source
add `convert` in layer : fonction to convert data from source to layer's elements

The  layer does
```js
const layer = {
  id, 
  type,
  source,
  convert,
}
```
The  provider does
``` js
function executeCommand() {
      const extent = command.extent; //  the datas's extent  to fetch
      const dataFetchedFromSource = await fetcher(source.urlFromExtent(extent));
      const dataParsed = await parse(dataFetchedFromSource);
      const layerElement = convert(dataParsed);
      return layerElement;
}
```
## Motivation
 * Isolate **`sources`** and **`converters/parsers`** from providers and use them for other layers 
 * simplify providers and merge certain providers (WMTS, WMS, Rasterizer), see delete them

mbredif commented on src/Core/TileMesh.js:
```diff
@@ -205,8 +204,9 @@ TileMesh.prototype.changeSequenceLayers = function changeSequenceLayers(sequence
 
 TileMesh.prototype.getCoordsForLayer = function getCoordsForLayer(layer) {
     if (layer.protocol.indexOf('wmts') == 0) {
```
layer.source.protocol ?

mbredif commented on src/Core/View.js:
```diff
@@ -177,12 +184,15 @@ function _preprocessLayer(view, layer, provider, parentLayer) {
 
     // probably not the best place to do this
     if (layer.type == 'color') {
+        layer.fx = layer.fx || 0.0;
```
why not a defineLayerProperty as for other properties?

zarov replied:
`defineLayerProperty` is a bad practice imho, and achieves nothing besides reducing a little the duplicated code, executing a callback in really rare usecases and emitting events on properties change. But I wouldn't change it here, and in the meantime keep this property declaration as it is.

mbredif reviewed (commented):
It looks better than the current state IMHO.
Some remarks after a quick review : 
- `layer.source || layer` looks odd, is it temporary ? can't we get rid of it ?
- converts and parsers are 2 different things (eg: WMTS source -> Geojson parser -> Mesh converter)

I however feel that tiledCRS version of extents should be an implementation detail of quadtree-based sources and that other parts of itowns should only handle extents as a (BBox3 + CRS).

Likewise instead of relative pitches / offsetscales, we could use absolute coordinates (BBox2 + CRS) for textures so that downscaling a texture would be a noop and uvs in the shader would be well defined coordinates with a proper crs instead of normalized coordinates.

zarov commented on src/Parser/textureConverter.js:
```diff
@@ -0,0 +1,20 @@
+};
+
+export default {
+    textureColorLayer(texture, extent, layer) {
```
I wouldn't call this textureColorLayer. Why not have two methods, `color` and `elevation`, which gives `textureConverter.color` and `textureConverter.elevation` calls.

Also by introducing this you could drop the `getColor/XBilTextureByUrl` in `OGCWebServiceHelper` which have nothing to do with OGC.

zarov commented on src/Provider/Fetcher.js:
```diff
@@ -1,14 +1,24 @@
-import { TextureLoader } from 'three';
+import { TextureLoader, DataTexture, AlphaFormat, FloatType, LinearFilter } from 'three';
```
My opinion: we should always import `* as THREE`, I don't get the appeal behind selecting what we are importing.

zarov commented on src/Provider/Fetcher.js:
```diff
@@ -90,10 +99,12 @@ export default {
-            checkResponse(response);
-            return response.arrayBuffer();
+    arrayBuffer,
+    textureFloat(url, options = {}) {
```
This part should be in the `elevationTexture` part imho. Let's keep the `Fetcher` simple.

zarov commented on src/Core/Scheduler/Scheduler.js:
```diff
@@ -282,9 +282,9 @@ Scheduler.prototype.addProtocolProvider = function addProtocolProvider(protocol,
-    if (typeof (provider.preprocessDataLayer) !== 'function') {
-        throw new Error(`Can't add provider for ${protocol}: missing a preprocessDataLayer function.`);
-    }
+    // if (typeof (provider.preprocessDataLayer) !== 'function') {
```
:+1:

zarov commented on src/Core/View.js:
```diff
@@ -139,17 +142,22 @@ function _preprocessLayer(view, layer, provider, parentLayer) {
+
+    // temp
+    // devrait etre supprime et placer dans source
+    if (provider && !layer.source) {
```
Yep, a provider shouldn't be the one defining the limit.

zarov commented on src/Core/View.js:
```diff
@@ -166,6 +174,15 @@ function _preprocessLayer(view, layer, provider, parentLayer) {
             if (!(providerPreprocessing && providerPreprocessing.then)) {
                 providerPreprocessing = Promise.resolve();
             }
+        } else if (layer.source) {
```
I don't like having this here, as much as I don't like having all other things below (conditions on type of layer). I know though that you hope to move it to the layer itself, so I'm ok with this here for now.

zarov reviewed (commented):
I think this is a good idea. The introduction of a `Source` object (kind of like OpenLayers ?) is a good step towards decluttering providers. Could you add a more accurate description (maybe documentation like the current `Provider` interface ?) to this object ?

However I don't really like a few things:
- the fetching in `geoFileSource`: no operation like should happen in a `Source`
- the `convert` method in `layer`: what about your initials thoughts with `source, processing, style` only ?

peppsac commented:
I'm all for refactoring this part, and the chosen solution looks similar to what was previously discussed.
(the only downside is that we already have almost 30 open PR and that updating/reviewing/merging others would be more relevant)

Random notes:
- IMHO, processing can depend on the data type (points or textures) but should not depend on the source type
- provider (or source) implements the protocol specific knowledge and don't necesarly need to do network accesses themselves
- a big refactoring must like this one:
  * not lose features (obviously)
  * bring new features (not necessarly in the same PR) to show its usefulness.

The last point is important. For instance, if this PR doesn't help implement vector-tiles support easily, then something is wrong.

nosy-b commented:
Sounds like the way to go!

zarov commented on index.html:
```diff
@@ -49,14 +49,15 @@
             promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/OrthosCRS.json').then(addLayerCb));
-            promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/ScanEX.json').then(addLayerCb));
+            // FIXME
+            // promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/ScanEX.json').then(addLayerCb));
```
What is the problem here ?

gchoqueux replied:
We can add all the layers without breaking the globe. I will fix it

zarov commented on src/Core/Geographic/Extent.js:
```diff
@@ -68,11 +68,15 @@ Extent.prototype.clone = function clone() {
 };
 
+Extent.prototype.isTiledCrs = function fnIsTiledCrs() {
+    return _isTiledCRS(this._crs);
```
You could move the code of `_isTiledCRS` directly here and call everywhere else in the file `this.isTiledCRS` only.

gchoqueux replied:
ok

zarov commented on src/Core/Scheduler/Scheduler.js:
```diff
@@ -132,21 +127,13 @@ Scheduler.prototype.constructor = Scheduler;
 
 Scheduler.prototype.runCommand = function runCommand(command, queue, executingCounterUpToDate) {
-    var provider = this.providers[command.layer.protocol];
+    var provider = this.getProtocolProvider(command.layer.protocol || command.layer.source.protocol);
```
Why move to `getProtocolProvider` ?

gchoqueux replied:
Encapsulation was already existing, it is better to use it. 
We can return a default value with this fonction

zarov commented on src/Core/TileMesh.js:
```diff
@@ -141,12 +141,12 @@ TileMesh.prototype.updateGeometricError = function updateGeometricError() {
     }
     if (textures) {
-        this.material.setTexturesLayer(textures, layerType, layerId);
+        this.material.setTexturesLayer(textures, layerType, layerId, pitchs);
```
What is the usage of `setTexturesLayer` ? Isn't it meant to be `setLayerTextures` ?

gchoqueux replied:
`Mesh.setTexturesLayer` and `this.material.setTexturesLayer` doesn't exist anymore

zarov commented on src/Core/TileMesh.js:
```diff
@@ -172,41 +172,42 @@ TileMesh.prototype.changeSequenceLayers = function changeSequenceLayers(sequence
+            return OGCWebServiceHelper.computeTMSCoordinates(this, source.extent, source.origin);
         }
-    } else if (layer.extent.crs() == this.extent.crs()) {
+        // FIXME: source.extent == undefined
```
What is the fix you have in mind ?

gchoqueux replied:
I revert this part, it's unnecessary

zarov commented on src/Parser/textureConverter.js:
```diff
@@ -0,0 +1,54 @@
+            texture = data;
+        }
+
+        if (layer.type === 'color') {
```
What about putting this part in `ColorLayer` and `ElevationTexture` ?

gchoqueux replied:
I don't like the actual textureConverter. It do too much things. I wan't add THREE.js code in layer. I could delete the layer parameter and have a more generic fonction.

zarov commented on src/Parser/GeoJsonParser.js:
```diff
@@ -211,6 +211,7 @@ function readFeatures(crsIn, crsOut, features, filteringExtent, options) {
             res.features.push(f);
         }
     }
+    res.isFeature = true;
```
:+1:

zarov commented on src/Layer/TiledGeometryLayer.js:
```diff
@@ -123,6 +124,31 @@ class TiledGeometryLayer extends GeometryLayer {
+        return convertToTile.convert(requester, extent, this);
+    }
+
+    buildRootTile(view, scheduler) {
```
Should be in the constructor imho.

gchoqueux replied:
yes, I move it in constructor

zarov commented on src/Parser/VectorTileParser.js:
```diff
@@ -4,17 +4,18 @@ import GeoJsonParser from './GeoJsonParser';
 
     if (layers.length < 1) return;
 
+    const crsInId = Number(options.crsIn.slice(5));
```
I don't understand the point of this. How can you be sure you got the correct value ?

gchoqueux replied:
the crs is always `EPSG:XXXX`, and there's 5 characters in `EPSG:`

zarov commented on src/Process/FeatureProcessing.js:
```diff
@@ -44,12 +55,15 @@ export default {
         }
 
         const features = node.children.filter(n => n.layer == layer);
+        // FIXME: traverse is do for each frame in each object3D
```
It is indeed a big problem, and could be a partial response to #731 thanks for discovering this.

zarov commented on src/Process/LayeredMaterialNodeProcessing.js:
```diff
@@ -326,29 +349,42 @@ export function updateLayeredMaterialNodeElevation(context, layer, node, parent)
+        (textures) => {
+            const terrain = { texture: textures[0] };
+
+            // deja verfier dans le provider
```
So does it mean you are going to remove things here ?

gchoqueux replied:
this check is maybe useless, see in `dataSourceProvider`

zarov replied:
It seems indeed. Maybe let it as it is for now and we'll see to remove that when refactoring this processing.

zarov commented on src/Provider/DataSourceProvider.js:
```diff
@@ -0,0 +1,135 @@
+import Cache from '../Core/Scheduler/Cache';
+import CancelledCommandException from '../Core/Scheduler/CancelledCommandException';
+
+const supportedFetchers = new Map([
```
Maybe we could agree on a default fetcher ? The default parser being none, I suggest that `Fetcher.texture` to be the default.

gchoqueux replied:
ok, how would you do this?

zarov replied:
Something simple like adding a `['default', Fetcher.texture]` entry to this map, and a simple check in the `fetchData` below. Even if you don't add a default fetcher, you need to catch the exception for a not supported format, as there seems to be no safe guard for now.

gchoqueux replied:
I do something for not supported format

zarov commented on src/Provider/DataSourceProvider.js:
```diff
@@ -0,0 +1,135 @@
+    [true, noParsingNeeded],
+]);
+
+// Todo: move this method to layer
```
Why ?

gchoqueux replied:
this function should be layer's callback in function of data

zarov commented on src/Provider/DataSourceProvider.js:
```diff
@@ -0,0 +1,135 @@
+            });
+}
+
+function extentToString(extent) {
```
Should be a `toString` method of `Extent`.

gchoqueux replied:
ok

zarov commented on src/Source/WMSSource.js:
```diff
@@ -0,0 +1,76 @@
+import Source from './Source';
+import URLBuilder from '../Provider/URLBuilder';
+
+const supportedFormats = [
```
Is this still relevant ?

gchoqueux replied:
I moved this part in `Source`. But, I think it's not respond to the question.
could you precise?

zarov replied:
No indeed, I was thinking that we could get rid of this array ? Even if the WMS spec says that files should be images, I think we can get rid of it.

gchoqueux replied:
removed

zarov commented on src/Source/fileSource.js:
```diff
@@ -0,0 +1,88 @@
+    return new Extent('EPSG:4326', -180, 180, -90, 90);
+}
+
+// TODO move and refacto
```
I'm OK to let it as it is for now, I have plan for this part.

gchoqueux replied:
what is your plan?

zarov replied:
Drop this function and add some parser, kinda like in #766

zarov commented on test/unit/layeredmaterialnodeprocessing.js:
```diff
@@ -28,18 +28,23 @@ describe('updateLayeredMaterialNodeImagery', function () {
-        layer.tileInsideLimit = () => true;
-        layer.visible = true;
+
         layer.updateStrategy = STRATEGY_MIN_NETWORK_TRAFFIC;
```
Side note, but this should be `layer.updateStrategy = { type: STRATEGY_MIN_NETWORK_TRAFFIC }`.

gchoqueux replied:
yes

zarov reviewed (changes requested):
I couldn't leave a comment on the file: the screenshot `wfs_color` is not used, please remove it.

Otherwise it looks good, we are heading a path I like. I saw a lot of `TODO`, what are your plans here ? Could you list the one you would do later (ie in another PR) ?

zarov commented:
I tested the examples, here's what I saw:
- __cubic_planar__: there seems to have artifact on the sides, see those two comparisons screenshots (first actual version, second this PR) 
![image](https://user-images.githubusercontent.com/741255/44451596-7e0a2f00-a5f4-11e8-971d-2cb861e5a802.png)
![image](https://user-images.githubusercontent.com/741255/44451567-6e8ae600-a5f4-11e8-882d-e9d37e9206c5.png)
- __globe_vector__: picking is not working anymore
- __syncCameras__: not working, I have a `Cannot read property 'protocol' of undefined`

gchoqueux commented:
@zarov all examples are fixed

New change:
 - root tile are build in constructor
 - `Layer.projection` are determined in `addLayer`

zarov commented on src/Source/Source.js:
```diff
@@ -0,0 +1,104 @@
+ */
+class Source {
+    /**
+     * @param  {sourceParams}  source  object to set source
```
You should remove the `@class` declaration above, and write a `@constructor` here, to avoid duplication in the generated page of the documentation. 

You should also write an example of declaring a random Source. Best would be to have it for all Sources files.

gchoqueux committed:
test(core): add new source examples

zarov commented on src/Source/FileSource.js:
```diff
@@ -0,0 +1,100 @@
+     */
+    constructor(source, crsOut) {
+        super(source);
+        if (!source.url) {
```
No need to repeat this as it is already been checked in `Source`.

gchoqueux replied:
yes

zarov commented on src/Source/FileSource.js:
```diff
@@ -0,0 +1,100 @@
+            throw new Error('source.url is required in FileSource');
+        }
+
+        if (!source.projection) {
```
You should specify in the documentation above that the projection is mandatory for this type of Source. This check can also be moved before calling `super(source)`.

zarov commented on src/Source/Source.js:
```diff
@@ -0,0 +1,104 @@
+     * @param      {Extent}  extent  extent to convert in url
+     * @return     {string}  url from extent
+     */
+    // eslint-disable-next-line
```
This is a problem I encountered as well, there is three solutions:
- `eslint-disable-next-line`
- declaring this method as `static`: `static urlFromExtent(extent)`, that can still be called with `this.urlFromExtent(extent)` or `source.urlFromExtent(extent)`
- removing the `class-method-use-this` eslint check from the configuration

gchoqueux replied:
yes, there are errors lint too for jsdoc

zarov commented on src/Source/StaticSource.js:
```diff
@@ -0,0 +1,131 @@
+    constructor(source) {
+        super(source);
+
+        if (!source.extent) {
```
Same as above, you can move this check before calling the parent constructor and specify in the documentation that the `extent` is required here.

zarov commented on src/Source/WFSSource.js:
```diff
@@ -0,0 +1,70 @@
+     * @param {Object} [source.zoom]
+     * @param {number} [source.zoom.min] layer's zoom minimum
+     * @param {number} [source.zoom.max] layer's zoom maximum
+     * @param {string} [source.zoom.max] layer's zoom maximum
```
Did you mean something else ?

zarov commented on src/Source/WFSSource.js:
```diff
@@ -0,0 +1,70 @@
+    */
+    constructor(source) {
+        super(source);
+        if (!source.typeName) {
```
Same, move the `super` below

zarov commented on src/Source/WMSSource.js:
```diff
@@ -0,0 +1,86 @@
+            }
+        }
+
+        let crsPropName = 'SRS';
```
Why not directly `const crsPropName = (this.version === '1.3.0') ? 'CRS' : 'SRS';` ?

zarov commented on src/Source/WMTSSource.js:
```diff
@@ -0,0 +1,100 @@
+
+/**
+ * Options to wtms protocol
+ * @typedef {Object} OptionsWmts
```
Will it be detailed ?

zarov reviewed (commented):
Added some comments on the Sources files !

gchoqueux committed:
test(core): add unit test on providers and sources

gchoqueux committed:
examples(core): add Source examples

gchoqueux committed:
feat(protocol): add Source

Add new object source:
 - Source have the properties to fetch the data to display
 - Remove all source abstractions from provider

gchoqueux commented:
@zarov PR updated

zarov reviewed (approved):
OK let's validate this, I think I didn't see any problem.

For fixing your failing test, either remove it, or install a polyfill for `URL` (like this one https://github.com/webcomponents/URL) as it is not a component supported in nodejs.

zarov commented:
It is even occurring because the `URL` component has only been added in node 10.

gchoqueux committed:
test(core): add source unit test
